### ISSUE TITLE
Support for Kotlin Native

### ISSUE THREAD

**Issue by spencerelliott** [google/filament#1197]
**Describe the solution you'd like**
Adding support for Kotlin Native would allow developers to create a single codebase that supports iOS, Desktop, and Android. The code would be compiled to native code for each platform possibly increasing performance.

**Describe alternatives you've considered**
It is currently possible to build macOS, PC, Android, and Linux apps using Kotlin. These platforms each require their own build to be performed, though. Apps for iOS need to be rewritten in C++.

**Comment by bejado** [comment]
I haven't explored Kotlin Native too deeply, but I think it's is an interesting idea. Not sure how much additional work it would be given we already have JNI bindings.

**Comment by romainguy** [comment]
What support do you need added exactly?

**Comment by spencerelliott** [comment]
The [`cinterop`](https://kotlinlang.org/docs/reference/native/c_interop.html) tool would need to be run on the Filament codebase. This would generate an interface between the native code and Kotlin Native. From there, I believe work would have to be done to build a stable API for each platform to pass in a valid surface using the [`expect/actual` mechanism](https://kotlinlang.org/docs/reference/platform-specific-declarations.html).

Additionally, it would be nice to have it available as a Multiplatform library through Maven. [Kotlin Native has support for publishing artifacts](https://kotlinlang.org/docs/reference/building-mpp-with-gradle.html#publishing-a-multiplatform-library). This may be a separate ticket, though.

**Comment by honza-vanek** [comment]
Filament is such a great piece of work; this would make it a killer lib. Any news on the issue? We are currently starting a rewrite of an AR app running on iOS tu support Android and multiplatform kotlin-native is the way to go. The possibility to use filament would be fantastic

**Comment by romainguy** [comment]
I agree supporting Kotlin Native would be fantastic. Unfortunately it is not a priority at this point. To do this properly we first need to find a scalable and sustainable way to support all the different language bindings since this would at least require to recreate the Java APIs in Kotlin.

That said a possible workaround for you right now would be to start from our JNI binding as a base for a C API for Filament. This should be usable with Kotlin.

**Comment by unicomp21** [comment]
FYI
https://kotlinlang.org/docs/reference/native/c_interop.html

**Comment by honza-vanek** [comment]
Thank you for the fast answers, sorry to hear that, will try to find some other way...
Good luck with filament and thanks for all the work

**Comment by chris-hatton** [comment]
...another developer here whose cross-platform 3D dreams have been stopped dead in their tracks by the realisation that Kotlin/Native can't currently use Filament due to incompatibility with C-interop tool.  Filament is a fabulous piece of work, would be a great 'force multiplier' to have these bindings available so it can be used as a Kotlin/Multiplatform library,

**Comment by honza-vanek** [comment]
Actually nope, they haven't. We have simply written a cross-platform renderer ourselves. Half a year delay but it's done :)

**Comment by romainguy** [comment]
@chris-hatton I understand but we just don't have resources to spare on this at the moment :(

**Comment by Erkko68** [comment]
Since this issue still comes up from time to time, I thought I'd share something I've been working on: https://github.com/Erkko68/filament-kmp.

It's a Kotlin Multiplatform implementation of Filament with an optional Compose abstraction layer. If Compose isn't your thing, the multiplatform packages can also be used directly, in a way that's similar to the Android Filament artifacts published on Maven.

I'd be happy to hear any feedback from people interested in using Filament with Kotlin Multiplatform.
